src/lambda_function/raw/app.py

```diff
@@ -0,0 +1,337 @@
+"""
+Raw Lambda
+
+A module for an AWS Lambda function that compresses JSON data contained in an
+export (zip archive) from S3 and uploads it to the raw S3 bucket. This module
+makes heavy use of Python file objects and multipart uploads and can
+decompress/compress/upload with a relatively low, fixed memory overhead
+with respect to the size of the uncompressed JSON.
+
+Example Usage:
+The module is intended to be deployed as an AWS Lambda function that listens to
+events from the `dispatch-to-raw` SQS queue.
+
+Environment Variables:
+- `RAW_S3_BUCKET`: The raw S3 bucket name where compressed data will be stored.
+- `RAW_S3_KEY_PREFIX`: The S3 key prefix within the raw bucket where data is written.
+"""
+
+import gzip
+import io
+import json
+import logging
+import os
+import zipfile
+
+import boto3
+
+logger = logging.getLogger()
+logger.setLevel(logging.INFO)
+
+
+def lambda_handler(event: dict, context: dict):
+    """
+    Entrypoint for this Lambda.
+
+    Args:
+        event (dict) An SQS message from the dispatch-to-raw SQS queue.
+        context (dict): Information about the runtime environment and
+            the current invocation
+    """
+    s3_client = boto3.client("s3")
+    raw_bucket = os.environ.get("RAW_S3_BUCKET")
+    raw_key_prefix = os.environ.get("RAW_S3_KEY_PREFIX")
+    main(
+        event=event,
+        s3_client=s3_client,
+        raw_bucket=raw_bucket,
+        raw_key_prefix=raw_key_prefix,
+    )
+
+
+def construct_raw_key(path: str, key: str, raw_key_prefix: str):
+    """
+    Constructs an S3 key for data to be uploaded to the raw S3 bucket.
+
+    Args:
+        path (str): The relative file path of the JSON data within its zip archive
+        key (str): The S3 key of the export/zip archive, formatted as `{namespace}/{cohort}/{export_basename}`.
+        raw_key_prefix (str): The raw S3 bucket key prefix  where all raw data is written.
+
+    Returns:
+        str: An S3 key in the format:
+            `{raw_key_prefix}/dataset={data_type}/cohort={cohort}/{basename}.ndjson.gz`
+
+    Notes:
+        - The function expects the input `key` to be formatted as `{namespace}/{cohort}/{export_basename}`.
+        - The data type is derived from the first underscore-delimited component of the file basename.
+    """
+    key_components = key.split("/")
+    # input bucket keys are formatted like `{namespace}/{cohort}/{export_basename}`
```

Comment:
Just checking, but is it always formatted like this?

Reply:
@thomasyu888 It could change in the future, but that would also mean changing every other stack that tries to read data from the input bucket, so no harm in hard coding this format.

Reply:
Lets hope it doesn't change in the future.